p2p/peer_tracker.go

```diff
@@ -60,15 +61,66 @@ func newPeerTracker(
 	return &peerTracker{
 		host:              h,
 		connGater:         connGater,
-		trackedPeers:      make(map[peer.ID]*peerStat),
-		disconnectedPeers: make(map[peer.ID]*peerStat),
+		trackedPeers:      make(map[libpeer.ID]*peerStat),
+		disconnectedPeers: make(map[libpeer.ID]*peerStat),
 		pidstore:          pidstore,
 		ctx:               ctx,
 		cancel:            cancel,
 		done:              make(chan struct{}, 2),
 	}
 }
 
+// bootstrap will bootstrap the peerTracker with the given trusted peers and will
+// block on the attempt to connect to `waitForNumConns` number of peers if the pidstore
+// returns enough previously-seen peers.
+//
+// NOTE: bootstrap is intended to be used with an on-disk peerstore.Peerstore as
+// the peerTracker needs access to the previously-seen peers' AddrInfo on start.
+func (p *peerTracker) bootstrap(ctx context.Context, trusted []libpeer.ID, waitForNumConns int) error {
+	// bootstrap connections to trusted
+	for _, trust := range trusted {
+		go p.connectToPeer(ctx, trust)
+	}
+
+	// short-circuit if pidstore was not provided
+	if p.pidstore == nil {
+		return nil
+	}
+
+	prevSeen, err := p.pidstore.Load(ctx)
+	if err != nil {
+		return err
+	}
+
+	// if not enough previously-seen peers were returned, just spawn
+	// all connection attempts async.
+	if len(prevSeen) < waitForNumConns {
+		waitForNumConns = 0
+	}
+	for i, peer := range prevSeen {
+		if i < waitForNumConns {
+			// we block on only the attempt to connect to the peer (successful or not)
+			p.connectToPeer(ctx, peer)
+			continue
+		}
```

Comment:
The idea behind blocking is to allow bootstrap to at least attempt N connections so that the likelihood peerTracker being empty / not filled enough during a call on the exchange (Head) is lower since Head can be called on node start.

Reply:
Why not make Head blocking instead?

Reply:
Okay the solution is included in https://github.com/celestiaorg/go-header/pull/53/commits/d4042e05d0d18311989a1d0239b455363bc4f9da

Instead of fully blocking, I just block on attempting to connect to trusted peers (regardless of err), and then spawn all other attempts to prev seen asynchronously which decreases the chances of head request happening on empty tracker.